vm/devices/storage/disk_nvme/nvme_driver/src/queues.rs

```diff
@@ -118,6 +143,28 @@ impl CompletionQueue {
             self.committed_head = self.head;
         }
     }
+
+    /// Saves queue data for servicing.
+    pub fn save(&self) -> CompletionQueueSavedState {
+        CompletionQueueSavedState {
+            cqid: self.cqid,
+            head: self.head,
+            committed_head: self.committed_head,
+            len: self.len,
+            phase: self.phase,
+            pfns: self.mem.pfns().to_vec(),
+        }
+    }
+
+    /// Restores queue data after servicing.
+    pub fn restore(&mut self, saved_state: &CompletionQueueSavedState) -> anyhow::Result<()> {
+        self.cqid = saved_state.cqid;
```

Comment:
We should not be overwriting fields that were initialized in `new`. Either `restore` should return `Self`, or it should only restore fields that were modified.

Additionally, we should be using the destructuring pattern to make sure we don't forget any fields:

```
let CompletionQueueSavedState {head, committed_head, phase} = self;
self.committed_head = committed_head;
...
```

Reply:
Looks like I need to redesign queue init a bit to satisfy returning Self from restore function.

Reply:
Refactored restore function to create an object instead of updating one.